crates/consensus/src/engine/mod.rs

```diff
@@ -0,0 +1,295 @@
+use futures::StreamExt;
+use reth_interfaces::consensus::ForkchoiceState;
+use reth_primitives::{
+    proofs::{self, EMPTY_LIST_HASH},
+    rpc::BlockId,
+    BlockLocked, Header, TransactionSigned, H64,
+};
+use reth_provider::{BlockProvider, HeaderProvider};
+use reth_rlp::Decodable;
+use reth_rpc_types::engine::{
+    ExecutionPayload, ForkchoiceUpdated, PayloadAttributes, PayloadStatus, PayloadStatusEnum,
+    TransitionConfiguration,
+};
+use std::{
+    collections::HashMap,
+    future::Future,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+use tokio::sync::oneshot;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+mod error;
+pub use error::EngineApiError;
+
+use crate::Config;
+
+/// The Engine API result type
+pub type EngineApiResult<Ok> = Result<Ok, EngineApiError>;
+
+/// The Engine API response sender
+pub type EngineApiSender<Ok> = oneshot::Sender<EngineApiResult<Ok>>;
+
+/// Consensus engine API trait.
+pub trait ConsensusEngine {
+    /// Retrieves payload from local cache.
+    fn get_payload(&self, payload_id: H64) -> Option<ExecutionPayload>;
+
+    /// Receives a payload to validate and execute.
+    fn new_payload(&mut self, payload: ExecutionPayload) -> EngineApiResult<PayloadStatus>;
+
+    /// Updates the fork choice state
+    fn fork_choice_updated(
+        &self,
+        fork_choice_state: ForkchoiceState,
+        payload_attributes: Option<PayloadAttributes>,
+    ) -> EngineApiResult<ForkchoiceUpdated>;
+
+    /// Verifies the transition configuration between execution and consensus clients.
+    fn exchange_transition_configuration(
+        &self,
+        config: TransitionConfiguration,
+    ) -> EngineApiResult<TransitionConfiguration>;
+}
+
+/// Message type for communicating with [EthConsensusEngine]
+pub enum EngineMessage {
+    /// New payload message
+    NewPayload(ExecutionPayload, EngineApiSender<PayloadStatus>),
+    /// Get payload message
+    GetPayload(H64, EngineApiSender<ExecutionPayload>),
+    /// Forkchoice updated message
+    ForkchoiceUpdated(
+        ForkchoiceState,
+        Option<PayloadAttributes>,
+        EngineApiSender<ForkchoiceUpdated>,
+    ),
+    /// Exchange transition configuration message
+    ExchangeTransitionConfiguration(
+        TransitionConfiguration,
+        EngineApiSender<TransitionConfiguration>,
+    ),
+}
+
+/// The consensus engine API implementation
+pub struct EthConsensusEngine<Client> {
```

Comment:
let's mark with with must_use

```suggestion
#[must_use = "EthConsensusEngine does nothing unless polled."]
pub struct EthConsensusEngine<Client> {
```